satysfi.opam

```diff
@@ -10,6 +9,7 @@ dev-repo: "git+https://github.com/gfngfn/SATySFi.git"
 bug-reports: "https://github.com/gfngfn/SATySFi/issues"
 build: [
   ["mkdir" "-p" "temp"]
+  ["dune" "subst"] {pinned}
```

Comment:
I'm curious. Why is this line executed only during pinning, but not installing non-pinned versions?

Reply:
I confirmed `%%VERSION%%` was not substituted when it was installed as a non-pinned package.

```
# Adding a package pointing to `git+https://github.com/nekketsuuu/SATySFi.git#68f8de124503634b05608f993b693dcf792d437b` to a local repository
$ opam install satysfi
$ satysfi --version
  SATySFi version %%VERSION%%
```

Reply:
Please read https://github.com/gfngfn/SATySFi/pull/167#issue-257002956

> This is intended; I think the version string is confusing and not needed for the development period. If needed, use `opam pin add`.

Reply:
Oops, Ioverlooked the paragraph. Sorry.  So, this PR changes the release process as well: `dune-release tag` instead of `git tag -a`. It will be helpful if the new release process is explicitly described somewhere in the repo.  Just my two cents.

Reply:
No, this pull request doesn't change the release process of SATySFi. Just running `git tag -a` is enough.

Reply:
This is because SATySFi is now installed/built using opam pin: https://github.com/gfngfn/SATySFi/blob/2271124d38ebd53d942e0cd05a3edfa958928e15/README.md#build

Reply:
Well, there are non-negligible number of users explicitly and implicitly installing SATySFi from an opam repo, namely [satyrographos-repo](https://github.com/na4zagin3/satyrographos-repo), including https://hub.docker.com/r/amutake/satysfi (pulled 7.8k times).  Although I can patch this line, I'd like not to add inessential patches.

BTW, I was wrong about `dune-release tag` which in fact does not replace the string.  It's required to run `dune-release` to upload a tarball as release asset.

Didn't you recommended @gfngfn adds SATySFi to the official OPAM repo?  I'd leave a notice somewhere so that he doesn't fall into the trap then.  Should I file an issue instead of mentioning it in the code base?

Reply:
> Didn't you recommended @gfngfn adds SATySFi to the official OPAM repo?

This is out of scope of this pull req. This pull request just makes it easy to confirm which version of SATySFi is used if HEAD is not tagged (i.e. not released) by adding its commit hash to the version string.

Reply:
@na4zagin3 Is there any proposed changes from you? Note that if I remove `{pinned}`, then `%%VERSION%%` is replaced every time after building SATySFi for development, and this process creates an unnecessary diff which should not be staged.

Reply:
It sounds totally reasonable to me to leave the line as is (i.e., with `{pinned}`); I'd like to have notes on the release process _somewhere,_ for example, a section in the README, or a new file:

> ### Release Process
> SATySFi uses dune-subst to substitute watermarks. Therefore it must be released in the following process so that prepared distributions (e.g., `satysfi-v0.0.6.tbz`) are uploaded to GitHub.
> 
> ```
> $ git tag add -v <version>
> $ dune-release publish distrib # or something like that
> ```

Reply:
Then it's out of scope of this pull request. I don't intend to change the way to release a new version. `dune-release` is not fit to current SATySFi.

Reply:
It _is_ a consequence of this PR, which changes how SATySFi should be released.

As OPAM repos prefers a source tarball in `url` section rather than source repos (especially, the official OPAM repo _prohibits_ source repos there), there are only two ways to substitute watermarks:

1. Release a prepared distribution, in other words, source code preprocessed by dune-subst. 
2. Upload a OPAM file which has a line like `["find" "." "-iname" "*.ml" "-exec" "sed" "-i.bak" "-e" "s/%%VERSION%%/v0.0.6/" "{}" ";"]` instead of `["dune" "subst"] {pinned}`.

Do you think “`dune-release` is not fit to current SATySFi” because it automatically attempts to release SATySFi to GitHub and to send a PR to the official OPAM repository?

I edited my previous comment with `dune-release distrib` that only generates a prepared distribution.  Does that make sense to you?

Reply:
Hmm, even though it is a consequence, it can be handled by another issue/PR.  You're right.  Shall I file one?

Reply:
I filed https://github.com/gfngfn/SATySFi/issues/237 since I don't want to block you on this. Today I learned a lot.  Thank you for a fruitful discussion!

Reply:
I don't think this pull req changes the current release process, though.... I commented my opinion to that issue <https://github.com/gfngfn/SATySFi/issues/237#issuecomment-660766570>. Let's continue this discussion there.